fix: fix issues seen in "multi" mode when a large number of sockets are created concurrently

gkc opened the pull request:
**DRAFT** until socket_connector 2.3.3 has been published and taken up by this branch

**- What I did**
fix: fix issues seen in "multi" mode when a large number of sockets are created concurrently

**- How I did it**
- added a missing trim() in srv_impl
- take up v2.3.3 of socket_connector which
  - ensures that `serverToSocket` introduces no delay in completing the tcp handshake
  - ensures all sockets' `done` future is handled upon its completion whether with or without error
  - ensures that `serverToSocket` handles sockets in strict sequence as they are accepted
  - ensures that, when one side is closed, all data received from that side has been delivered to the other side before closing the other side
- test: added a feature to sshnpd whereby if `INLINE_SRV` is in the environment and set to `true`, then we run the Srv inline rather than forking it in a separate process. This is super useful for debugging srv issues on the daemon side.

**- How to verify it**
Manual testing required. I've tested as follows with local atDirectory and atServers:
- Run policy admin UI, which listens on port 3000
`build/sshnp/np_admin -a @policy -n sshnp --root-domain vip.ve.atsign.zone`
- Run noports daemon allowing port 3000 
`bin/sshnpd --root-domain vip.ve.atsign.zone -m @client -a @service -d mbp --permit-open "*:*"`
- Run srvd locally
`bin/srvd -a @relay --root-domain vip.ve.atsign.zone --ip 127.0.0.1 -v`
- Run npt requesting remote port 3000 and local forwarding port 3001
`bin/npt --root-domain vip.ve.atsign.zone -f @client -t @service -r @relay --rh 127.0.0.1 --rp 3000 -d mbp --lp 3001 -T 0`
- open http://localhost:3001
- Reload the page many times

gkc committed:
fix: fix issues seen in "multi" mode when a large number of sockets are created concurrently
- added a missing trim() in srv_impl
- take up v2.3.0 of socket_connector which ensures that `serverToSocket` introduces no delay in completing the tcp handshake

xavierchanth commented:
Some stress testing, the improvement is solid enough for most webpages. For a single webpage to create enough connections to overload this version means that webpage is probably doing something very wrong.

gkc committed:
Merge branch 'trunk' into fix-multi-highly-concurrent

gkc commented:
Once https://github.com/atsign-foundation/socket_connector/pull/21 has been merged and published then I will bump the dependency in this PR and mark it ready for review

xavierchanth commented:
e2e test failures are my fault, I accidentally merged a bad commit, will resolve this shortly.

xavierchanth committed:
Merge branch 'trunk' into fix-multi-highly-concurrent

gkc committed:
build: take up socket_connector 2.3.1
fix: as good practice, added calls to `await socket.flush()` everywhere we're writing stuff to sockets in srv_impl.dart
test: added a feature to sshnpd whereby if `INLINE_SRV` is in the environment and set to `true`, then we run the Srv inline rather than forking it in a separate process. This is super useful for debugging srv issues on the daemon side.

gkc committed:
Merge remote-tracking branch 'origin/fix-multi-highly-concurrent' into fix-multi-highly-concurrent

gkc committed:
Merge branch 'trunk' into fix-multi-highly-concurrent

gkc commented:
Took up some latest socket_connector package and made some more tweaks to noports_core (srv_impl.dart). Updated PR description accordingly

gkc committed:
chore: removed some commented-out code

gkc committed:
Merge remote-tracking branch 'origin/fix-multi-highly-concurrent' into fix-multi-highly-concurrent

gkc committed:
fix: fix race condition introduced by previous commits

gkc committed:
fix: fix race condition introduced by previous commits

gkc committed:
fix: take up socket_connector release candidate 2.3.2

gkc committed:
Merge remote-tracking branch 'origin/trunk' into fix-multi-highly-concurrent

gkc committed:
build: take up socket_connector 2.3.2; nump noports_core package version

gkc committed:
build: bump noports_core version in sshnoports' pubspec

gkc committed:
build: bumping sshnoports version to 5.6.4

gkc committed:
build: use socket_connector 2.3.3 from git branch

gkc committed:
feat: if "NP_TRACE" is in the environment as 'true' then set verbose:true when calling SocketConnector.serverToSocket

gkc committed:
feat: stuff to help troubleshooting
- if "SRV_TRACE" is in the environment as 'true' then set verbose:true when calling SocketConnector.serverToSocket
- if "SRV_INLINE" is in the environment as 'true' then run the srv inside the sshnpd process rather than forking

gkc committed:
build: take up next candidate 2.3.3 socket_connector

gkc committed:
test: add a couple of simple programmes to help with testing where we want to shove lots of data around

gkc committed:
test: add a couple of simple programmes to help with testing where we want to shove lots of data around

gkc committed:
build: pubspec.lock update

gkc committed:
fix: npt: gracefully handle failures to connect to 'local' hosts / ports on the daemon side. Previously the entire npt session would end; now the client side just sees a socket closing immediately
refactor: some refactoring for readability

gkc committed:
chore: dart format

gkc committed:
build: update pubspecs to pull latest commit from socket_connector fix-2.3.3 branch

gkc commented:
@gkc, @cconstab and @XavierChanth have successfully stress tested with various websites and concurrent load tests; this PR is functionally complete and ready to go.

Next steps:
- merge https://github.com/atsign-foundation/socket_connector/pull/25 and publish socket_connector 2.3.3
- on this PR, update pubspecs in noports_core and sshnoports to use socket_connector 2.3.3
- review, merge to trunk
- publish new noports release candidate

gkc commented on packages/dart/noports_core/lib/src/srvd/socket_connector.dart:
```diff
@@ -92,14 +92,15 @@ void socketConnector(ConnectorParams connectorParams) async {
+    logTraffic: logTraffic,
+    socketAuthVerifierA: socketAuthVerifierA,
+    socketAuthVerifierB: socketAuthVerifierB,
+  );
```
format change only

gkc commented on packages/dart/noports_core/lib/src/sshnpd/sshnpd_impl.dart:
```diff
@@ -596,21 +597,37 @@ class SshnpdImpl implements Sshnpd {
-        timeout: req.timeout,
-      ).run();
-      logger.info('Started rv - pid is ${rv.pid}');
+      if (Platform.environment['SRV_INLINE'] == 'true') {
```
Feature very useful for debugging sshnpd where it will run the srv in the same process rather than forking it

gkc commented on packages/dart/sshnoports/bin/demo_socket_client.dart:
```diff
@@ -0,0 +1,33 @@
+/// - simple client which talks to a demo_socket_server
+/// - sends a request for N kBytes of data to be sent to us
+/// - listens for the data and writes a message at the end
+/// saying how much it received vs expected
```
added this and demo_socket_server.dart to facilitate highly concurrent load testing of npt

gkc commented on packages/dart/noports_core/lib/src/srv/srv_impl.dart:
```diff
@@ -406,7 +406,11 @@ class SrvImplDart implements Srv<SocketConnector> {
+      if (relayAddresses.isEmpty) {
+        throw Exception('Cannot resolve relay host $streamingHost');
+      }
+      InternetAddress relayAddress = relayAddresses[0];
```
refactor: get the relayAddress at the start, and pass it down through the various functions

gkc commented on packages/dart/noports_core/lib/src/srv/srv_impl.dart:
```diff
@@ -472,15 +479,15 @@ class SrvImplDart implements Srv<SocketConnector> {
+      addressB: relayAddress,
       portB: streamingPort,
-      verbose: false,
+      verbose: Platform.environment['SRV_TRACE'] == 'true',
```
useful debugging feature to make the socket connector emit more logs

gkc commented on packages/dart/noports_core/lib/src/srv/srv_impl.dart:
```diff
@@ -472,15 +479,15 @@ class SrvImplDart implements Srv<SocketConnector> {
       multi: multi,
       timeout: timeout,
-      beforeJoining: (Side sideA, Side sideB) async {
+      beforeJoining: (Side sideA, Side sideB) {
```
has to be synchronous to avoid a race in `SocketConnector.handleSingleConnection`

gkc commented on packages/dart/noports_core/lib/src/srv/srv_impl.dart:
```diff
@@ -630,69 +637,111 @@ class SrvImplDart implements Srv<SocketConnector> {
   }
 
-  Future _handleMultiConnectRequest(
+  Future<void> _handleMultiConnectRequest(
```
refactor - `_handleMultiConnectRequest` is now just handling a multi connect request; the parsing of what was received on the control socket is now done in the new `_handleControlRequest` function

gkc commented on packages/dart/noports_core/lib/src/srv/srv_impl.dart:
```diff
@@ -630,69 +637,111 @@ class SrvImplDart implements Srv<SocketConnector> {
+    SocketConnector sc,
+    InternetAddress relayAddress,
+    DataTransformer? encrypter,
+    DataTransformer? decrypter,
```
refactor: Passing nullable encrypter and decrypter so we don't have two blocks of duplicate code for handling unencrypted vs encrypted connections

gkc commented on packages/dart/noports_core/lib/src/srv/srv_impl.dart:
```diff
@@ -630,69 +637,111 @@ class SrvImplDart implements Srv<SocketConnector> {
+        ' creating new socketToSocket connection');
+
+    InternetAddress localAddress = await resolveRequestedLocalHost();
+
```
In the code below, we're taking direct control of creating the socket connections (to the relay and to the requested local `host:port`) rather than calling `SocketConnector.socketToSocket`.

This is required because (1) the order really matters here (connect to relay first, connect to local `host:port` second) (2) write and flush the auth string to the relay before trying to create the local socket (3) ensure that there are no uncaught exceptions on either socket

gkc commented on packages/dart/noports_core/lib/src/srv/srv_impl.dart:
```diff
@@ -630,69 +637,111 @@ class SrvImplDart implements Srv<SocketConnector> {
+    InternetAddress localAddress = await resolveRequestedLocalHost();
+
+    // First, connect to the relay.
+    // If it fails, return, as we can't do anything more
```
Note that once we have connected and authenticated to the relay, the relay will join it to the client-side.

gkc commented on packages/dart/noports_core/lib/src/srv/srv_impl.dart:
```diff
@@ -630,69 +637,111 @@ class SrvImplDart implements Srv<SocketConnector> {
+    }
+
+    // Now, connect to the local host:port
+    // If it fails, we need to close the socket connection to the relay
```
When we close the socket connection to the relay, the relay will close its connection to the client.

gkc commented on packages/dart/noports_core/lib/src/srv/srv_impl.dart:
```diff
@@ -630,69 +637,111 @@ class SrvImplDart implements Srv<SocketConnector> {
+    unawaited(sc.handleSingleConnection(sideA).catchError((err) {
+      logger.severe(
+          'ERROR $err from handleSingleConnection on sideA (local - $localAddress:$localPort)');
+    }));
```
Once we've added both sides, socket connector will join them up so traffic starts to flow from local socket to relay socket (and within the relay, from this side's socket to the client side's socket, and on the client side from its relay socket to its local socket.)

gkc commented on packages/dart/noports_core/lib/src/srv/srv_impl.dart:
```diff
@@ -792,29 +843,49 @@ class SrvImplDart implements Srv<SocketConnector> {
     }
   }
 
+  Future<InternetAddress> resolveRequestedLocalHost() async {
```
- `InternetAddress.lookup`'s `type` parameter defaults to `InternetAddressType.any`
- Per that method's documentation, the order of the returned list is not guaranteed. 
- We will try to resolve via IPv4 first; if that doesn't work then we will try to resolve via IPv6
- We have more work to do in this area, since ipv6 addresses contain colons and we have code scattered around which parses things based on colons. In general we are not properly supporting IPv6. I've created [a ticket](https://github.com/atsign-foundation/noports/issues/1537) for remedying that

gkc commented:
@gkc, @cconstab and @XavierChanth have successfully stress tested with various websites and concurrent load tests; this PR is functionally complete and ready to go.

Next steps:
- [x] merge https://github.com/atsign-foundation/socket_connector/pull/25 and publish socket_connector 2.3.3
- [x] on this PR, update pubspecs in noports_core and sshnoports to use socket_connector 2.3.3
- [x] review, merge to trunk
- [ ] (**IN PROGRESS**) publish new noports release candidate
- [ ] deploy new srvd to the `@stream` relay
- [ ] retest https://github.com/atsign-foundation/noports/issues/1476
- [ ] publish new noports release
- [ ] deploy newly published srvd to rv_am, rv_eu, rv_ap etc

gkc committed:
build: update pubspecs to use socket_connector 2.3.3

gkc committed:
Merge remote-tracking branch 'origin/trunk' into fix-multi-highly-concurrent
